Add VariantTable component and related files

hextraza opened the pull request:
#### Rationale
Implements basic case of the variant data grid per https://github.com/BimberLab/DiscvrLabKeyModules/issues/119

Commit:
Add VariantTable component and related files

bbimber commented on jbrowse/src/client/JBrowse/VariantTable/VariantTable.tsx:
```diff
@@ -0,0 +1,105 @@
+                }
+
+                // Parse multiple location strings, separated by colons
+                const locStrings = locStringsRaw.split(":")
```
Two things: first, splitting on colon doesnt work with JBrowse's location formatting (i.e. Contig:Start..Stop). However, I think we should drop the requirement to support multiple regions as once, since this will make it easiest to toggle back and forth between the Browser and Table widgets. I think if you look further upstream of this code anyway, like line 44, that would break with an array of locations too.

bbimber commented on jbrowse/src/client/JBrowse/VariantTable/components/Buttons.tsx:
```diff
@@ -0,0 +1,29 @@
+    );
+  }
+
+export function FilterButton(props) {
```
For these filters, we should do one of two things:

- Read the column's data type. If numeric, dont allow filtering. The default 'contains' type filter doesnt really make sense for numbers.
- Alternately, we could create a generalizable two-field component to select operator (GT, LT, EQ) and value. This is not a huge feature, but would need to be fairly complete, including: 1) handling of situations where the user gives incomplete input (like operator without value). 
- The default 'contains' style filtering that exists now is perfectly fine for string columns.

Remember that the VCF header (I believe this is accessible under feature.metadata), should have a dictionary of fields, which includes the data type for each.

bbimber commented on jbrowse/src/client/JBrowse/VariantTable/components/VariantTableWidget.tsx:
```diff
@@ -0,0 +1,184 @@
+  }
+
+  // Ensure you can't arrow-navigate out of a filter input box when you're typing into it
+  function inputStopPropagation(event: React.KeyboardEvent<HTMLInputElement>) {
```
Like we discussed, it would be useful to give this table a more feature-rich header. This should include:
- The RefNameAutoComplete component, which drives the location for this table. If the URL has a location, we should populate this field from that. If there is no location, just load with no rows. The user should then be able to type into the box and load different areas (including the existing gene-search).
- Should move the export buttons from the bottom to this new button bar 
- Add either a dropdown menu or buttons for 'Filter By Attributes', and 'Filter by Sample'. These are existing options from ExtendedVariantPlugin. These are each a modal dialog that accepts the track object, and save properties to the Display config. In theory, this will cause the adapter to automatically filter, and our table should magically react to this.

bbimber replied:
There should also be a button for 'View in Genome Browser', which just page-navigates and loads that same session with the same location in view.

bbimber commented on jbrowse/src/client/JBrowse/VariantTable/components/VariantTableWidget.tsx:
```diff
@@ -0,0 +1,184 @@
+
+  const columns = [
+    {
+      key: 'id',
```
Should remove the ID column from the set of visible columns

bbimber commented on jbrowse/src/client/JBrowse/VariantTable/components/VariantTableWidget.tsx:
```diff
@@ -0,0 +1,184 @@
+  } else {
+    const gridElement = (
+      <>
+        <DataGrid
```
Consider this a reach goal, but it would be useful to the user if the grid had some kind of details icon, and if you click this it would open the same ExtendedVariantWidget modal that loads when one clicks a variant on the genome browser.

I suspect figuring out how to integrate this VariantTable widget and the other filter UI dialogs will take some work. However, if you solve this problem there, I would hope doing a very similar thing for the details view would be possible.

bbimber commented on jbrowse/src/client/JBrowse/VariantTable/dataUtils.ts:
```diff
@@ -0,0 +1,102 @@
+// Logic behind how to filter each field.
+export function filterFeature(feature, filters) {
+    return (
+        (filters.chrom ? feature.chrom.includes(filters.chrom) : true) &&
```
I didnt thoroughly check this codepath, but can any of properties be null? I suggest loading this with the session 'demo', trackid=clinvar_ncbi_hg38

hextraza replied:
The whole codepath should be handled for null/undefined -- none of the fields in filters should ever be able to be null

Commit:
Add RefNameAutocomplete and minor changes (still needs CSS for the toolbar spacing)

Commit:
Add SampleFilterWidget component

Commit:
Add InfoFilterWidget and dialog for widgets

Commit:
Swap from dataAdapter to RPC for getFeatures call

Commit:
Add working filtration and CSS changes

Commit:
Fix key errors when opening the filtration UI

Commit:
Style header cell height conditional on filtration on or off

Commit:
Add filtration test and change order/look of toolbar buttons

Commit:
Re-add all other tests

bbimber committed:
Test fix

bbimber commented on jbrowse/src/client/JBrowse/VariantTable/VariantTable.tsx:
```diff
@@ -0,0 +1,104 @@
+    const [rpcManager, setRpcManager] = useState(null)
+    const [assembly, setAssembly] = useState(null)
+
+    function generateViewState(genome){
```
Could we pull all or most of this code that queries LabKey for the session and builds the createViewState() into one piece of helper code, shared across Browser, StandaloneSearch and VariantTable? The reason is that this code is getting more complex, and this would give us two important things: 1) browser sets up some custom JBrowse themes that are not happening here, 2) browser reads some URL params not happening here. See Browser.tsx, which is probably the most complete codepath and should be used as the base.

bbimber commented on jbrowse/src/client/JBrowse/VariantTable/components/VariantTableWidget.tsx:
```diff
@@ -0,0 +1,285 @@
+        exportToXlsx(gridElement, 'rows.xlsx')
+        break;
+      case "browserRedirect":
+        window.location.href = ActionURL.buildURL("jbrowse", "jbrowse.view", null, {session: sessionId, location: locString, trackId: trackId})
```
Can you pull this into a separate util function? The reason is that we should serialize the INFO and Sample filters on the URL. We should use shared code here and in:

[ExtendedVariantDisplay/model.ts](https://github.com/BimberLab/DiscvrLabKeyModules/pull/146/files#diff-4c4bb7ef784f5abc8f0a87bfe4c76a8e08dbc28410f85cdbc99e54fe82dcbf0b)

This way we always have consistent behavior when jumping from browser<->table.

bbimber replied:
see Browser.tsx for parsing of ActiveSamples and InfoFilters from the URL. As suggested above, we should centralize the code that calls getSession and generateViewState() into some shared helper, which will help this

bbimber reviewed (changes requested):
This is really close. The two main things are:

1) We should ensure we serialize/deserialize all the filters on the URL when jumping to/from browser/table. Because this code is getting more complex, I suggest we pull all the URL-parsing and URL-construction code into common helpers, and use these for Browser, StandAloneSearch, and VariantTable

2) We need to make the theme consistent. See screenshot below for example of this widget if the site these was changed. I believe centralizing the getSession/generateViewState code, using Browser as the base, will help. That has code that sets the JBrowse theme based on the site. 

2B) If we're lucky getting the JBrowse theme will fix this; however, note that the info/sample filter modals dont look quite right. They should inherit the JBrowse theme.

2C) you might need to set the classes of your button bar buttons to mirror however the JBRowse theme is set up

3) Can you think about whether you should add <ErrorBoundary> to the VariantTable, like Browser.tsx? This basically catches error and presents them better. 

![image](https://user-images.githubusercontent.com/7724863/162579772-88d2eec9-902d-466e-b3e9-56e108e3dc8f.png)

maybe setting the theme will fix this, but can this button be made wider so the text fits on one line?

![image](https://user-images.githubusercontent.com/7724863/162579810-ee6048bf-bd91-4ec8-af42-5ff72b7b2a57.png)

bbimber committed:
Add column in VariantTable to show variant type

bbimber committed:
Test fix

Commit:
Add ThemeProvider and centralize the getSession and url redirection code in uitils, and parse SampleFilters from the URL

Note: the InfoFilters are still broken as of this patch, because the track we're getting from the fetchSession function doesn't have an InfoFilters field on it

Commit:
Merge branch '21.11_fb_variantdatagrid' of github.com:BimberLab/DiscvrLabKeyModules into 21.11_fb_variantdatagrid

Commit:
Add InfoFilter field to the track when deserializing from URL

Commit:
Add error handling to the fetch/util callbacks and allow the table to load on an empty locstring

Commit:
Increase maxRegionSize and add try/catch around tests that throw an alert
